### PR TITLE
TCG-2460 - Automatically refresh frequently used keys before they expire

### PR DESCRIPTION
**Pull request by guywithnose**
#### What does this PR do?
TCG-2460

#### Checklist
- [  ] Pull request contains a clear definition of changes
- [  ] Tests (either unit, integration, or acceptance) written and passing
- [  ] Relevant documentation produced and/or updated

### CODE REVIEW

File: src/Predis.php [hunk 1]
```diff
@@ -51,23 +60,82 @@ public function memoizeCallable(string $key, callable $compute, int $cacheTime =
     {
         if (!$this->refresh) {
             try {
-                $cached = $this->client->get($key);
+                if (rand(1, 100) <= $this->refreshPercent) {
```
Comment (chrisryan) [inline src/Predis.php]: Should this support checking for 0 if someone never wants to use this functionality?

File: src/Predis.php [hunk 2]
```diff
@@ -51,23 +60,82 @@ public function memoizeCallable(string $key, callable $compute, int $cacheTime =
...
+                    // {$refreshPercent}% of requests should check to see if this key is almost expired.
+                    // We don't want to check this on every call to preserve performance.
+                    // Also, this functionality is only important for requests that have many concurrent calls.
+                    $runtime = $this->client->get("{$key}.runtime");
+                    $ttl = $this->client->pttl($key) / 1000;
+                    if ($runtime && $runtime * 3 > $ttl) {
```
Comment (chrisryan) [inline src/Predis.php]: Should there be a way to change the multiplier as a config?

File: src/Predis.php [hunk 3]
```diff
@@ -51,23 +60,82 @@ public function memoizeCallable(string $key, callable $compute, int $cacheTime =
...
+            $data = json_decode($cached, true);
+            return $data['result'];
+        }
+    }
+
+    private function getLock($key)
+    {
+        $this->client->watch("{$key}.lock");
+        $lock = $this->client->get("{$key}.lock");
+        if ($lock !== null && $lock < microtime(true) - 3) {
```
Comment (chrisryan) [inline src/Predis.php]: Should this wait time be configurable as well?

**Comment by chadicus** [comment]
In my opinion the functionality you are adding should not be in the `Predis` class, but instead be part of a custom `ClientInterface` instance.

File: src/Predis.php (hunk 1 above)
Comment (guywithnose) [reply src/Predis.php]: If it's 0 then this will always be false

**Comment by guywithnose** [comment]
I pulled out the locking part, because as @chadicus pointed out that may work better at a different abstraction level.  The part that measures runtime still needs to be here, so I left that.

**Comment by guywithnose** [comment]
@chadicus Here is where I moved the locking part: https://github.com/traderinteractive/lib-predis-anti-stampede-php/pull/1

**Comment by guywithnose** [comment]
I set refreshPercent to default to 0.  That should make it fully BC

**Pull request merged**